### ISSUE TITLE
Applying an SVG filter to the canvas

### ISSUE THREAD

**Issue by velara3** [samizdatco/skia-canvas#157]
The [documentation](https://www.npmjs.com/package/skia-canvas) says, "Skia canvas supports the full set of CSS filter image processing operators"

In the examples I've seen applying a filter is as simple as setting the filter property:  

```
const canvas = document.getElementById("canvas");
const ctx = canvas.getContext("2d");

ctx.filter = "blur(4px)";
```

The CSS filter documentation mentions the URL() filter:  
<img width="722" alt="image" src="https://github.com/samizdatco/skia-canvas/assets/1084127/cb5f4e68-b527-4c5b-b5d6-6b9736fb3a9b">

> A CSS [url()](https://developer.mozilla.org/en-US/docs/Web/CSS/url). Takes an IRI pointing to an SVG filter element, which may be embedded in an external XML file.

How would I do the same in Skia Canvas?

**Comment by mpaperno** [comment]
There is currently no support for any SVG features at all, nor is there anything to handle `url()` types. The filters currently parsed can be found in [the code here](https://github.com/samizdatco/skia-canvas/blob/main/src/filter.rs#L74), for what it's worth... which I think is all of them _except_ `url()`.

Skia itself must support it, but I have no idea about _skia-safe_ (the intermediate library layer used by _skia-canvas_), or what would be involved in implementing it. Just being able to load SVGs natively would be a great start.

**Comment by velara3** [comment]
It looks like [Skia](https://en.wikipedia.org/wiki/Skia_Graphics_Engine) itself has SVG support. I couldn't find a library with the term `skia safe` but I see the import statements in the code and a few links to [Rust Skia](https://github.com/rust-skia/rust-skia). 

I don't know that I know how to implement it either. It would be great to have. 

Basically, I'm trying to apply filters to canvas elements server side and the supported filters are currently the CSS filters but they are somewhat primitive if IIUC. The URL() filter would allow [SVG filters](https://blog.logrocket.com/complete-guide-using-css-filters-svgs/) which IIUC offer much wider dynamic range. 

Although, if it was possible, having a way to apply multiple filters in code would also solve the issue. Something like:  

```
var filters = [];
var dropShadow = new DropShadow(3,3,1, rgba(0,0,0,.5));
filters.push(dropShadow);
var blur = new Blur(3,6);
filters.push(blur);
var displacementMap = new DisplacementMap();
filters.push(displacementMap);
canvas.filters = filters;
```

**Comment by mpaperno** [comment]
Yea, sorry, "Rust Skia" is what I meant (it's called "skia safe" in the Rust package manager for some reason).  And from what I understand it support most, if not all, Skia features. But _skia-canvas_ only implements a subset.  There's PR #155 which would probably be a good place to start on updating.

Perhaps I misunderstand what you mean, but multiple filters can be added now like in the [standard](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/filter#applying_multiple_filters), by separating the filter functions with spaces in the `ctx.filter =` call, eg. `ctx.filter = "contrast(1.4) sepia(1) drop-shadow(-9px 9px 3px #e81)"`.

Cheers,
-Max

**Comment by velara3** [comment]
I'm aware of adding multiple filters inline. What I meant by adding multiple filters manually was to do it longhand in JavaScript code rather than by string values. So instead of:  

```
ctx.filter = "contrast(1.4) sepia(1) drop-shadow(-9px 9px 3px #e81)";
```
something like this: 
```
var filters = [];
var aFoundationalFilter = new FoundationalFilter();
aFoundationalFilter.matrix = new Matrix(0,0,1,0,0,1);
aFoundationalFilter.offsetX = 10;
aFoundationalFilter.offsetY = 10;
filters.push(aFoundationalFilter);
var compositeFilter = new CompositeFilter();
compositeFilter.valueA = 20;
compositeFilter.valueB = 20;
filters.push(compositeFilter);
ctx.applyFilters(filters);
```

I'd rather create the filter objects and set values on them if that makes sense. Some of my past experience in image data filters make sense and some areas I've still got more to learn.

**Comment by mpaperno** [comment]
Well there's no "filter object type" in standard JS that I know of, and the context `filter` property only takes a string, so.... do be more imperative about adding filters like you suggest, you'd need to add those features somehow. Even if there already was a `applyFilters()` convenience method, currently it would still just be parsing filter strings, presumably... so I'm not sure what the benefit would be.

You could have your filter objects output a string as a result, collect them into an array (eg. `filters.push(compositeFilter.toString())`), then simply `ctx.filter = filters.join(' ')`. 

That's actually almost exactly what I've done in some code where the app's users can add arbitrary canvas filters as image layers.

**Comment by mpaperno** [comment]
Totally OT now, but if there was a built-in way to track added filters so that they can later be reset automatically, w/out having to specify the filter again with "no op" values (eg. `blur(0)`), that would be pretty slick. For example when one wants to apply filter(s) to a particular drawing layer/path within a larger composition.  Can be added on top as a JS layer of course, but having that in the binary core would likely be a lot more efficient.

**Comment by velara3** [comment]
Indeed. One would hope that there would be class references for all the HTML and SVG markup. So that they could be applied and modified like you describe.

**Comment by velara3** [comment]
If there's anything I can do to help push this along (besides writing the code - I'm unable to contribute at this moment) let me know.

**Comment by samizdatco** [comment]
Adding support for `url()` filters seems like it would be a pretty massive undertaking since it involves http & SVG-parsing even beyond the filter itself. `skia_safe` now has types related to SVG filter [elements](https://rust-skia.github.io/doc/skia_safe/svg/fe/index.html), but it's not clear to me how you go from those `Node` objects (which really just seem to represent the markup) to an equivalent set of Skia filters that can be used in rendering. 

If anyone would like to investigate this and suggest a path forward I'd be happy to pick it up, but in the meantime I don't feel like there's an urgent need given the ability to compose CSS filter chains using strings as @mpaperno describes above.
